src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferDictionaryContainsMethods.cs

```diff
@@ -0,0 +1,223 @@
+
+                    IMethodSymbol? containsKeyMethod = property.ContainingType.GetMembers(ContainsKeyMethodName)
+                        .OfType<IMethodSymbol>()
+                        .WhereAsArray(x => x.ReturnType.SpecialType == SpecialType.System_Boolean &&
+                            x.IsPublic() &&
+                            x.Parameters.Length == 1 &&
+                            x.Parameters[0].Type.Equals(keyType, SymbolEqualityComparer.Default))
+                        .SingleOrDefault();
+
+                    if (containsKeyMethod is null)
+                        return;
+
+                    var diagnostic = Diagnostic.Create(ContainsKeyRule, invocation.Syntax.GetLocation(), property.ContainingType.Name);
+                    context.ReportDiagnostic(diagnostic);
+                }
+                else if (property.Name == ValuesPropertyName)
+                {
+                    if (!containsParameterType.Equals(valueType, SymbolEqualityComparer.Default))
+                        return;
+
+                    //  Now we search for an accessible ContainsValue method that returns boolean and accepts a single
+                    //  parameter of the type that was substituted for TValue.
+
+                    IMethodSymbol? containsValueMethod = property.ContainingType.GetMembers(ContainsValueMethodName)
+                        .OfType<IMethodSymbol>()
+                        .WhereAsArray(x => x.ReturnType.SpecialType == SpecialType.System_Boolean &&
+                            x.Parameters.Length == 1 &&
+                            x.Parameters[0].Type.Equals(valueType, SymbolEqualityComparer.Default))
+                        .SingleOrDefault();
+
+                    if (containsValueMethod is null)
+                        return;
+
+                    var diagnostic = Diagnostic.Create(ContainsValueRule, invocation.Syntax.GetLocation(), property.ContainingType.Name);
+                    context.ReportDiagnostic(diagnostic);
+                }
+            }
+
+            static bool TryGetPropertyReferenceOperation(IInvocationOperation containsInvocation, [NotNullWhen(true)] out IPropertySymbol? property)
+            {
+                //  If the method being invoked is an extension method, then the first child may be an IArgumentOperation, 
+                //  and its first child may be an IConversionOperation. The first child is not guaranteed to be an IArgumentOperation
+                //  because VB and C# handle extension methods differently. Likewise, the IConversionOperation may be absent if
+                //  the compile-time types match exactly. Therefore, we need to walk down conditionally. 
+
+                IOperation current = containsInvocation;
+                if (containsInvocation.TargetMethod.IsExtensionMethod)
+                {
+                    if (current.Children.FirstOrDefault() is IArgumentOperation argumentOperation)
```

Comment:
It feels a little wrong to me to be doing all these `Children` calls.

Reply:
I've fixed it so it uses more derived operation types with explicitly exposed properties.